src/hotspot/share/opto/vtransform.cpp

```diff
@@ -697,69 +697,68 @@ bool VTransformGraph::has_store_to_load_forwarding_failure(const VLoopAnalyzer&
+}
+
+Node* VTransformApplyState::transformed_node(const VTransformNode* vtn) const {
+  Node* n = _vtnode_idx_to_transformed_node.at(vtn->_idx);
+  assert(n != nullptr, "must find IR node for vtnode");
   return n;
 }
 
-VTransformApplyResult VTransformScalarNode::apply(const VLoopAnalyzer& vloop_analyzer,
-                                                  const GrowableArray<Node*>& vnode_idx_to_transformed_node) const {
+VTransformApplyResult VTransformScalarNode::apply(VTransformApplyState& apply_state) const {
   // This was just wrapped. Now we simply unwap without touching the inputs.
   return VTransformApplyResult::make_scalar(_node);
 }
 
-VTransformApplyResult VTransformReplicateNode::apply(const VLoopAnalyzer& vloop_analyzer,
-                                                     const GrowableArray<Node*>& vnode_idx_to_transformed_node) const {
-  Node* val = find_transformed_input(1, vnode_idx_to_transformed_node);
+VTransformApplyResult VTransformReplicateNode::apply(VTransformApplyState& apply_state) const {
+  Node* val = apply_state.transformed_node(in_req(1));
   VectorNode* vn = VectorNode::scalar2vector(val, _vlen, _element_type);
-  register_new_node_from_vectorization(vloop_analyzer, vn, val);
+  register_new_node_from_vectorization(apply_state, vn, val);
   return VTransformApplyResult::make_vector(vn, _vlen, vn->length_in_bytes());
 }
 
-VTransformApplyResult VTransformConvI2LNode::apply(const VLoopAnalyzer& vloop_analyzer,
-                                                   const GrowableArray<Node*>& vnode_idx_to_transformed_node) const {
-  Node* val = find_transformed_input(1, vnode_idx_to_transformed_node);
+VTransformApplyResult VTransformConvI2LNode::apply(VTransformApplyState& apply_state) const {
+  Node* val = apply_state.transformed_node(in_req(1));
   Node* n = new ConvI2LNode(val);
-  register_new_node_from_vectorization(vloop_analyzer, n, val);
+  register_new_node_from_vectorization(apply_state, n, val);
   return VTransformApplyResult::make_scalar(n);
 }
 
-VTransformApplyResult VTransformShiftCountNode::apply(const VLoopAnalyzer& vloop_analyzer,
-                                                      const GrowableArray<Node*>& vnode_idx_to_transformed_node) const {
-  PhaseIdealLoop* phase = vloop_analyzer.vloop().phase();
-  Node* shift_count_in = find_transformed_input(1, vnode_idx_to_transformed_node);
+VTransformApplyResult VTransformShiftCountNode::apply(VTransformApplyState& apply_state) const {
+  PhaseIdealLoop* phase = apply_state.phase();
+  Node* shift_count_in = apply_state.transformed_node(in_req(1));
   assert(shift_count_in->bottom_type()->isa_int(), "int type only for shift count");
   // The shift_count_in would be automatically truncated to the lowest _mask
   // bits in a scalar shift operation. But vector shift does not truncate, so
   // we must apply the mask now.
   Node* shift_count_masked = new AndINode(shift_count_in, phase->igvn().intcon(_mask));
```

Comment:
Randomly noticed this: You should probably use `phase->intcon()` which also sets root as control`. Same at other places further down. You might want to squeeze that into this patch as well.

Reply:
I'm doing it in a few SuperWord related places now.